src/core/include/units/quantity_spec.h

```diff
@@ -0,0 +1,518 @@
+ * However, quantities of the same dimension are not necessarily of the same kind.
+ *
+ * The character of those quantities is derived from ingredients or overriden with a template parameter.
+ *
+ * User should derive a strong type from this class template rather than use it directly in the source code.
+ * For example:
+ *
+ * @code{.cpp}
+ * // quantity kinds
+ * inline constexpr struct width : quantity_spec<length> {} width;
+ * inline constexpr struct height : quantity_spec<length> {} height;
+ * inline constexpr struct thickness : quantity_spec<width> {} thickness;
+ * inline constexpr struct diameter : quantity_spec<width> {} diameter;
+ * inline constexpr struct position_vector : quantity_spec<length, quantity_character::vector> {} position_vector;
+ *
+ * // derived quantities
+ * inline constexpr struct area : quantity_spec<pow<2>(length)> {} area;
+ * inline constexpr struct volume : quantity_spec<pow<3>(length)> {} volume;
+ * inline constexpr struct velocity : quantity_spec<position_vector / duration> {} velocity;  // vector
+ * inline constexpr struct speed : quantity_spec<distance / duration> {} speed;
+ * inline constexpr struct force : quantity_spec<mass * acceleration, quantity_character::vector> {} force;
+ * inline constexpr struct power : quantity_spec<force * velocity, quantity_character::scalar> {} power;
+ * @endcode
+ *
+ * Two quantity specifications are deemed equal when they are of the same type. With that, both strong
+ * types `speed` and `velocity` are considered not equal to `derived_dimension<length, per<time>>` or
+ * to each other. However, they are both convertible to `derived_dimension<length, per<time>>`.
+ * User can implicitly convert up and down the inheritance hierarchy between them. However, `speed` and
+ * `velocity` are not convertible as they do not inherit from each other. They are from two separate branches
+ * of dimensionally equivalent quantities.
+ *
+ * @tparam Q quantity specification of a parent quantity
+ * @tparam Args optionally a value of a `quantity_character` in case the base quantity should not be scalar
+ */
+#ifdef __cpp_explicit_this_parameter
+template<QuantitySpec auto Q, one_of<quantity_character> auto... Args>
+struct quantity_spec<Q, Args...> : std::remove_const_t<decltype(Q)> {
+#else
+template<typename Self, QuantitySpec auto Q, one_of<quantity_character> auto... Args>
+struct quantity_spec<Self, Q, Args...> : std::remove_const_t<decltype(Q)> {
+#endif
+  static constexpr auto kind_of = Q;
+  static constexpr quantity_character character = detail::quantity_character_init<Args...>(Q.character);
+
+#ifndef __cpp_explicit_this_parameter
+  template<Unit U>
+  [[nodiscard]] constexpr auto operator[](U u) const
+    requires(this->dimension == detail::get_dimension_for(u))
+  {
+    return reference<Self{}, u>{};
+  }
+#endif
+};
+
+#ifdef __cpp_explicit_this_parameter
+
+#define QUANTITY_SPEC(name, ...)                                \
```

Comment:
If this is to be part of the public API, it should be in its own header for the sake of C++ module importers.

Reply:
This is mainly a workaround to make the definitions work before C++23 is provided. As this C++23 feature will be implemented in all compilers probably way sooner than the modules support 😜 I do not think it is an issue.